### PR TITLE
eksctl info command

### LINKED ISSUE #3029: eksctl info command
**Why do you want this feature?**
Currently we ask users to provide the output from `eksctl version` and `kubectl version` when opening a bug issue. I want users to be able to easily obtain all necessary information in one command.

**What feature/behavior/change do you want?**
A command `eksctl info` that provides the `eksctl version`, `kubectl version` and the OS type. This can be expanded to show more output that we deem useful. 

```bash
$ eksctl info
eksctl version: 0.35.0
kubectl version: 1.18.2
OS: MacOS
```

Thoughts @weaveworks/eksctl-reviewers ?

### PR DESCRIPTION
**Pull request by neha-viswanathan**
Closes #3029 

### Description

This MR contains an implementation of the `eksctl info` command

### Checklist
- [x] Added tests that cover your change (if possible)
- [x] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes
- [x] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Comment by neha-viswanathan** [comment]
@aclevername 👋🏽 this is a rough first attempt of the `eksctl info` command. Please review and let me know your suggestions.

**Comment by aclevername** [comment]
Thanks @neha-viswanathan ! Looks good, could we do some error handling? For example if we fail to discover the kubectl client we should just present `unknown` or something like that :smile:

**Comment by neha-viswanathan** [comment]
@aclevername ready for round 2 :)

**Comment by neha-viswanathan** [comment]
@Callisto13 Hi! I'll need your guidance here - I don't understand why the `make lint` is failing? Please help!

**Comment by Callisto13** [comment]
No problem 😃 .

So the linter is reporting this error:
```
errors in package: [/__w/eksctl/eksctl/cmd/eksctl/main.go:45:49: undeclared name: infoCmd]"
```

Because the `infoCmd` function you are referencing on line 45 of `cmd/eksctl/main.go` does not exist. (This line here: `cmdutils.AddResourceCmd(flagGrouping, rootCmd, infoCmd)`.) You need to create that `infoCmd` function, which should call your new `GetVersion` function.

You can use the line below in `cmd/eksctl/main.go` which wires up the `versionCmd` as a guide. `versionCmd` is defined in `cmd/eksctl/version.go`, so following that pattern you can create a new `cmd/eksctl/info.go` which will define `infoCmd` and call your `GetVersion` function.

Hope this helps, let me know if you have any more questions or if anything is unclear 👍

**Comment by neha-viswanathan** [comment]
@Callisto13 for some reason that file did not get committed earlier. 😕 Thanks for pointing it out. 😃

**Comment by Callisto13** [comment]
> for some reason that file did not get committed earlier

Gah! that's annoying 😅

File: cmd/eksctl/info.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	cmd.FlagSetGroup.InFlagSet("General", func(fs *pflag.FlagSet) {
+		fs.StringVarP(&output, "output", "o", "", "specifies the output format (valid option: json)")
+	})
+	cmd.CobraCommand.Args = cobra.NoArgs
+	cmd.CobraCommand.RunE = func(_ *cobra.Command, args []string) error {
+		switch output {
+		case "":
+			fmt.Printf("eksctl version: %s\n", info.GetVersion().EksctlVersion)
+			fmt.Printf("kubectl version: %s\n", info.GetVersion().KubectlVersion)
+			fmt.Printf("OS: %s\n", info.GetVersion().OS)
```
Comment (Callisto13) [inline cmd/eksctl/info.go:25]: Do you think we could just call `GetVersion` just the once in this block? Would be good to avoid execing to kubectl 3 times in a row?

File: pkg/info/info.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+// Version holds versions info
+type Version struct {
+	EksctlVersion  string
+	KubectlVersion string
+	OS             string
+}
+
+// GetVersion returns versions info
+func GetVersion() Version {
```
Comment (Callisto13) [inline pkg/info/info.go:24]: Is is possible to have some tests for this file please?

File: pkg/info/info.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	}
+	return values[len(values)-1]
+}
+
+// String returns eksctl info as a string
+func String() string {
+	if data, err := json.Marshal(GetVersion()); err == nil {
+		return string(data)
+	}
+	return ""
```
Comment (Callisto13) [inline pkg/info/info.go:54]: Can we do the inverse of this?
```suggestion
	if data, err := json.Marshal(GetVersion()); err != nil {
		return fmt.Sprintf("error : %v", err)
	}
	return string(data)
```

File: pkg/info/info.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+type Version struct {
+	EksctlVersion  string
+	KubectlVersion string
+	OS             string
+}
+
+// GetVersion returns versions info
+func GetVersion() Version {
+	return Version{
+		EksctlVersion:  version.String(),
```
Comment (Callisto13) [inline pkg/info/info.go:26]: This returns quite a lot of stuff:
```
eksctl version: {"Version":"0.44.0","PreReleaseID":"dev","Metadata":{"BuildDate":"2021-03-26T15:34:25Z","GitCommit":"ee316d40"},"EKSServerSupportedVersions":["1.15","1.16","1.17","1.18","1.19"]}
```


I also think whatever is happening here is messing with the json (escaping everything):
```
$ ./eksctl info -o json
{"EksctlVersion":"{\"Version\":\"0.44.0\",\"PreReleaseID\":\"dev\",\"Metadata\":{\"BuildDate\":\"2021-03-26T15:34:25Z\",\"GitCommit\":\"ee316d40\"},\"EKSServerSupportedVersions\":[\"1.15\",\"1.16\",\"1.17\",\"1.18\",\"1.19\"]}","KubectlVersion":"v1.19.4","OS":"linux"}
```

I think some tests would help drive out desired UX in both these cases.

**Review by Callisto13 — CHANGES_REQUESTED** [review]

File: pkg/info/info.go (hunk 4 above)
Comment (neha-viswanathan) [reply pkg/info/info.go:26]: @Callisto13 instead of returning all this info, would it be enough to just display the version?

Comment (Callisto13) [reply pkg/info/info.go:26]: yes, ideally we would just show `eksctl version: 0.40.0` or similar

Comment (neha-viswanathan) [reply pkg/info/info.go:26]: @Callisto13 calling GetVersion() for retrieving eksctl's version returns a different version (version set in pkg/version/release.go). So I think we should go back to the previous implementation I had by executing the `eksctl version` command. Thoughts?
 ```
./eksctl info  
eksctl version: 0.44.0-dev
kubectl version: v1.19.7
OS: darwin

eksctl version
0.42.0
```

Comment (Callisto13) [reply pkg/info/info.go:26]: I would rather avoid shelling out to ourselves if we don't have to. Can we add a func somewhere to get the exact thing we need internally?

**Comment by Callisto13** [comment]
Hey @neha-viswanathan! Just checking in on the status of this, do you plan to finish or would you like it taken over? Thanks!

**Comment by neha-viswanathan** [comment]
> Hey @neha-viswanathan! Just checking in on the status of this, do you plan to finish or would you like it taken over? Thanks!

Hi @Callisto13 I'll try and wrap this up within the next week or so. I will keep you posted in any case.

**Comment by Callisto13** [comment]
Awesome 🎉

File: pkg/info/info.go (hunk 4 above)
Comment (neha-viswanathan) [reply pkg/info/info.go]: @Callisto13 🙋🏽‍♀️ Need some pointers on how I can extract the versions internally? Thanks!

Comment (Callisto13) [reply pkg/info/info.go]: Hey! Just looking back and I realised I made a mistake 🤦‍♀️ : 

Above when you built and ran the command locally you got:

```sh
./eksctl info  
eksctl version: 0.44.0-dev
kubectl version: v1.19.7
OS: darwin
```

and thought the version was wrong as it used the version set in pkg/version/release.go. This was actually correct, since the version is a local dev build built from main, which will not be the same as if you just run `eksctl version`. using your installed release version.

So TLDR: you _can_ use `version.GetVersion()`, it will return the correct thing.

File: pkg/info/info_test.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	It("wrong eksctl, kubectl versions and OS", func() {
+		info := GetVersion()
+
+		Expect(info).To(Equal(Version{
+			EksctlVersion:  "0.40.0",
+			KubectlVersion: "v1.10.1",
+			OS:             "amd64",
+		}))
+	})
+
```
Comment (Callisto13) [inline pkg/info/info_test.go:39]: I think this test can be simpler. The `kubectl` and `OS` are very unlikely to be ignored, although I am sure you can find a way to force them to disappear. Likewise there will never be a "wrong" version for any of them.

```suggestion
 var _ = Describe("Info", func() {
   Context("GetInfo", func() {
     It("returns the eksctl and kubectl versions, and the host os", func() {
       result := info.GetInfo()

       Expect(result.EksctlVersion).ToNot(Equal(""))
       Expect(result.KubectlVersion).ToNot(Equal(""))
       // we can't check for a specific version since it will change depending
       // on who is running the tests, but we can check that actual semvers are returned
       _, err := semver.NewVersion(strings.TrimRight(result.EksctlVersion, "\n"))
       Expect(err).NotTo(HaveOccurred())
       _, err = semver.NewVersion(strings.TrimRight(result.KubectlVersion, "\n"))
       Expect(err).NotTo(HaveOccurred())

       Expect(result.OS).ToNot(Equal(""))
       // technically a cheat, but
       // testing for the OS is harder, we could possibly have an array of OSes and see
       // if the one here  matches, if you feel like doing that
       Expect(result.OS).To(Equal(runtime.GOOS))
     })
   })

   Context("String", func() {
     It("returns Info in json", func() {
       result := info.String()

       // make processing easier, turn into a struct
       versions := info.Versions{}
       err := json.Unmarshal([]byte(result), &versions)
       Expect(err).NotTo(HaveOccurred())

       // same again basically, we can just check they are not nil
       Expect(versions.EksctlVersion).ToNot(Equal(""))
       Expect(versions.KubectlVersion).ToNot(Equal(""))
       Expect(versions.OS).ToNot(Equal(""))
     })
   })
   
   // There are is a potential failure case, but it is trickier
   // to test at this level. To test that the right thing happens when kubectl fails
   // to execute, you could remove the binary at the start of the test and put it back at the end
   // but if that goes wrong then you dont have a kubectl anymore :D 
   // Another option is to have the kubectl binary be an exported location which you can reset
   // in the test to a dummy bin. But this is a lot of work for a small functionality,
   // so if that sound like a pain I am happy to ignore it 
 })
```

Comment (neha-viswanathan) [reply pkg/info/info_test.go:39]: @Callisto13 thank you so much! Those tests definitely simplified everything.
1. regarding the OS test, it looks like these OSes are supported: https://github.com/golang/go/blob/master/src/go/build/syslist.go#L10. 
2. The kubectl test of moving the binary sounds a bit of an overkill. 

I feel the tests you have suggested look great. And thanks again for helping me out.

File: pkg/info/info.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	return values[len(values)-1]
+}
+
+// String return info as JSON
+func String() string {
+	if data, err := json.Marshal(GetInfo()); err == nil {
+		return string(data)
+	}
+	return ""
+}
```
Comment (Callisto13) [inline pkg/info/info.go:60]: let's print something here when the marshal fails, just so users don't get confused about why the command returned nothing:

```suggestion
func String() string {
	data, err := json.Marshal(GetInfo())
	if err != nil {
		return fmt.Sprintf("failed to marshal info into json: %q", err)
	}
	return string(data)
}
```

File: pkg/info/info_test.go (hunk 5 above)
Comment (Callisto13) [reply pkg/info/info_test.go]: > regarding the OS test, it looks like these OSes are supported: https://github.com/golang/go/blob/master/src/go/build/syslist.go#L10.

cool, if you want to implement a match loop there feel free. 

> The kubectl test of moving the binary sounds a bit extra.

👍  no worries

File: pkg/info/info_test.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+			Expect(result.OS).ToNot(Equal(""))
+
+			_, err := semver.NewVersion(strings.TrimSpace(result.EksctlVersion))
+			Expect(err).NotTo(HaveOccurred())
+
+			_, err = semver.NewVersion(strings.TrimSpace(strings.TrimPrefix(result.KubectlVersion, "v")))
+			Expect(err).NotTo(HaveOccurred())
+
+			oses := []string{"aix", "android", "darwin", "dragonfly", "freebsd", "hurd", "illumos", "ios", "js", "linux", "nacl", "netbsd", "openbsd", "plan9", "solaris", "windows", "zos"}
+			Expect(result.OS).To(BeElementOf(oses))
```
Comment (Callisto13) [inline pkg/info/info_test.go:28]: oooh i didn't know `BeElementOf` was a thing! nice!

**Review by Callisto13 — APPROVED** [review]
LGTM 🚀  thanks for working on this!
